# 🧑‍💼 📣 Brienfing do gerente

##### “Danyllo, agora que você tem os dados carregados no data mart SQLite, preciso que você me entregue uma análise inicial das vendas de 2015 a 2017.

#### Quero entender rapidamente:

#### Como estão distribuídas as vendas?
#### Quem são nossos melhores clientes?
#### Quais produtos puxam a receita?
#### Como as vendas evoluem ao longo do ano?

#### Me entregue isso via SQL primeiro, antes de irmos para o dashboard.”

# Objetivo da Etapa 1

✅ Conectar ao banco SQLite

✅ Listas as tabelas

✅ Entender os campos

✅ Criar tabelas unificadas (se necessário)

✅ Executar consultas de negócio

## Script Oficial - Projeto Vendas

#### 1. Fontes dos CSVs:

* Repositório com tabelas (produto, cliente, calendário etc.)
  https://github.com/cnduta/Azure_AdventureWorks_DE_pipeline/tree/main/data/Raw

* Repositório com arquivos CSV completos (tabelas de produção e vendas)
  https://github.com/olafusimichael/AdventureWorksCSV


In [1]:
import pandas as pd
import sqlite3
import requests
import io

#-------------------------------
# 1) URLs dos CSVs oficiais
#-------------------------------

urls = {
    "Calendar": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Calendar.csv",
    "Customers": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Customers.csv",
    "Products": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Products.csv",
    "Sales_2015": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Sales_2015.csv",
    "Sales_2016": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Sales_2016.csv",
    "Sales_2017": "https://raw.githubusercontent.com/cnduta/Azure_AdventureWorks_DE_pipeline/main/data/Raw/AdventureWorks_Sales_2017.csv"
}

#-------------------------------
# 2) Criando bando SQLite
#-------------------------------

conn = sqlite3.connect("adventureworks.db")

print("Baixando e importando tabelas...")

for table, url in urls.items():
    print(f"Importando {table}...")
    response = requests.get(url)
    df = pd.read_csv(io.StringIO(response.text))
    df.to_sql(table, conn, if_exists="replace", index=False)

conn.close()

print("Banco adventuresworks.db criado com sucesso!")

Baixando e importando tabelas...
Importando Calendar...
Importando Customers...
Importando Products...
Importando Sales_2015...
Importando Sales_2016...
Importando Sales_2017...
Banco adventuresworks.db criado com sucesso!


## 1. Conexão ao SQLite

In [2]:
import sqlite3

conn = sqlite3.connect("adventureworks.db")
cursor = conn.cursor()

# Teste simples
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()
tables

[('Calendar',),
 ('Customers',),
 ('Products',),
 ('Sales_2015',),
 ('Sales_2016',),
 ('Sales_2017',)]

## 2. Vendo estrutura de cada tabela

In [3]:
cursor.execute("PRAGMA table_info(Products);")
cursor.fetchall()

[(0, 'ProductKey', 'INTEGER', 0, None, 0),
 (1, 'ProductSubcategoryKey', 'INTEGER', 0, None, 0),
 (2, 'ProductSKU', 'TEXT', 0, None, 0),
 (3, 'ProductName', 'TEXT', 0, None, 0),
 (4, 'ModelName', 'TEXT', 0, None, 0),
 (5, 'ProductDescription', 'TEXT', 0, None, 0),
 (6, 'ProductColor', 'TEXT', 0, None, 0),
 (7, 'ProductSize', 'TEXT', 0, None, 0),
 (8, 'ProductStyle', 'TEXT', 0, None, 0),
 (9, 'ProductCost', 'REAL', 0, None, 0),
 (10, 'ProductPrice', 'REAL', 0, None, 0)]

In [4]:
cursor.execute("PRAGMA table_info(Calendar);")
cursor.fetchall()

[(0, 'Date', 'TEXT', 0, None, 0)]

In [5]:
cursor.execute("PRAGMA table_info(Customers);")
cursor.fetchall()

[(0, 'CustomerKey', 'INTEGER', 0, None, 0),
 (1, 'Prefix', 'TEXT', 0, None, 0),
 (2, 'FirstName', 'TEXT', 0, None, 0),
 (3, 'LastName', 'TEXT', 0, None, 0),
 (4, 'BirthDate', 'TEXT', 0, None, 0),
 (5, 'MaritalStatus', 'TEXT', 0, None, 0),
 (6, 'Gender', 'TEXT', 0, None, 0),
 (7, 'EmailAddress', 'TEXT', 0, None, 0),
 (8, 'AnnualIncome', 'TEXT', 0, None, 0),
 (9, 'TotalChildren', 'INTEGER', 0, None, 0),
 (10, 'EducationLevel', 'TEXT', 0, None, 0),
 (11, 'Occupation', 'TEXT', 0, None, 0),
 (12, 'HomeOwner', 'TEXT', 0, None, 0)]

In [6]:
cursor.execute("PRAGMA table_info(Sales_2015);")
cursor.fetchall()

[(0, 'OrderDate', 'TEXT', 0, None, 0),
 (1, 'StockDate', 'TEXT', 0, None, 0),
 (2, 'OrderNumber', 'TEXT', 0, None, 0),
 (3, 'ProductKey', 'INTEGER', 0, None, 0),
 (4, 'CustomerKey', 'INTEGER', 0, None, 0),
 (5, 'TerritoryKey', 'INTEGER', 0, None, 0),
 (6, 'OrderLineItem', 'INTEGER', 0, None, 0),
 (7, 'OrderQuantity', 'INTEGER', 0, None, 0)]

In [7]:
cursor.execute("PRAGMA table_info(Sales_2016);")
cursor.fetchall()

[(0, 'OrderDate', 'TEXT', 0, None, 0),
 (1, 'StockDate', 'TEXT', 0, None, 0),
 (2, 'OrderNumber', 'TEXT', 0, None, 0),
 (3, 'ProductKey', 'INTEGER', 0, None, 0),
 (4, 'CustomerKey', 'INTEGER', 0, None, 0),
 (5, 'TerritoryKey', 'INTEGER', 0, None, 0),
 (6, 'OrderLineItem', 'INTEGER', 0, None, 0),
 (7, 'OrderQuantity', 'INTEGER', 0, None, 0)]

In [8]:
cursor.execute("PRAGMA table_info(Sales_2017);")
cursor.fetchall()

[(0, 'OrderDate', 'TEXT', 0, None, 0),
 (1, 'StockDate', 'TEXT', 0, None, 0),
 (2, 'OrderNumber', 'TEXT', 0, None, 0),
 (3, 'ProductKey', 'INTEGER', 0, None, 0),
 (4, 'CustomerKey', 'INTEGER', 0, None, 0),
 (5, 'TerritoryKey', 'INTEGER', 0, None, 0),
 (6, 'OrderLineItem', 'INTEGER', 0, None, 0),
 (7, 'OrderQuantity', 'INTEGER', 0, None, 0)]

## 3. Criando tabela unificada de vendas (VIEW)

##### Para análises, vamos juntar as tabelas dos anos ed 2015+2016+2017 em uma unica tabela

In [9]:
cursor.execute("""
CREATE VIEW IF NOT EXISTS Sales_All AS
SELECT * FROM Sales_2015
UNION ALL
SELECT * FROM Sales_2016
UNION ALL
SELECT * FROM Sales_2017;
               """)
conn.commit()

## 4. Primeiras Queries

### 4.1 Total de vendas por ano

In [10]:
def run(query):
    cursor.execute(query)
    rows = cursor.fetchall()
    for row in rows:
        print(row)

In [11]:
run("""
SELECT
    strftime('%Y', OrderDate) AS Ano,
    SUM(OrderQuantity) AS Total_Qtd
FROM Sales_All
GROUP BY Ano
ORDER BY Ano;
""")

(None, 84174)


#### Porque o resultado deu "(None, 84174)"?

Significa que:
* Ano = None
* Tota_Qtd = 84174

Ou seja:

O SQLite não conseguiu extrair o ano da coluna "OrderDate", e por isso retornou None.
Mas conseguiu fazer a soma "SUM(OrderQuantity) e retornou 84.174 vendas.

Portanto:
✔ O banco está funcionando
✔ A view Sales_All está ok
✔ Os dados foram importados
❌ O formato da data não está reconhecido como data para o strftime()
Isso acontece muito quando o CSV tem datas no formato diferente do esperado.

##### Investigando a causa

Os CSVs do AdventureWorks (dependendo do repositório) podem vir assim:

* 1/1/2015
* 2015-01-01 00:00:00
* 2015-01-01
* 2015/01/01

Então primeiro precisamos ver um exemplo real da coluna OrderDate.

In [12]:
run("""
SELECT OrderDate
FROM Sales_ALL
LIMIT 10;
""")

('1/1/2015',)
('1/1/2015',)
('1/1/2015',)
('1/1/2015',)
('1/2/2015',)
('1/2/2015',)
('1/2/2015',)
('1/2/2015',)
('1/3/2015',)
('1/3/2015',)


##### Solução: Converter o formato da data

A coluna OrderDate está no formato M/D/YYYY, que é um formato texto e não é reconhecido pelo SQLite como data nativa.

SQLite só consegue usar strftime() quando a coluna está em um formato aceito como data:

 ✔ YYYY-MM-DD

 ✔ YYYY-MM-DD HH:MM:SS

 ✔ Formatos ISO

Esse formato M/D/YYYY não segue o padrão ISO, então o SQLite interpreta como texto comum, não data.

In [13]:
run("""
SELECT
    substr(OrderDate, -4) AS Ano,
    SUM(OrderQuantity) AS Total_Qtd
FROM Sales_All
GROUP BY Ano
ORDER By Ano;
""")

('2015', 2630)
('2016', 36230)
('2017', 45314)


##### Por que usar substr(OrderDate, -4)?

Porque:

O SQLite lê a data como texto
-4 pega os 4 últimos caracteres
Como o ano sempre tem 4 dígitos → funciona para qualquer M/D/YYYY

É uma técnica comum em ETL leve quando não queremos ainda transformar toda a coluna em data nativa.

### 4.2 Top 10 Produtos mais Vendidos

In [14]:
run("""
SELECT
    p.ProductKey,
    p.ProductName,
    SUM(s.OrderQuantity) AS TotalVendida
FROM Sales_All s
JOIN Products p ON p.ProductKey = s.ProductKey
GROUP BY p.ProductKey, p.ProductName
ORDER BY TotalVendida DESC
LIMIT 10;
""")

(477, 'Water Bottle - 30 oz.', 7967)
(480, 'Patch Kit/8 Patches', 5898)
(528, 'Mountain Tire Tube', 5678)
(529, 'Road Tire Tube', 4327)
(223, 'AWC Logo Cap', 4151)
(485, 'Fender Set - Mountain', 3960)
(478, 'Mountain Bottle Cage', 3810)
(479, 'Road Bottle Cage', 3329)
(530, 'Touring Tire Tube', 2740)
(536, 'ML Mountain Tire', 2119)


### 4.3 Top 10 Clientes (por quantidade comprada)

In [15]:
run("""
SELECT
  c.CustomerKey,
  c.FirstName || ' ' || c.LastName AS FullName,
  SUM(s.OrderQuantity) AS TotalComprado
FROM Sales_All s
JOIN Customers c ON c.CustomerKey = s.CustomerKey
GROUP BY c.CustomerKey, FullName
ORDER BY TotalComprado DESC
LIMIT 10;
""")

(11262, 'JENNIFER SIMMONS', 106)
(11300, 'FERNANDO BARNES', 106)
(11331, 'SAMANTHA JENKINS', 102)
(11185, 'ASHLEY HENDERSON', 100)
(11566, 'APRIL SHAN', 99)
(11091, 'DALTON PEREZ', 97)
(11277, 'CHARLES JACKSON', 97)
(11287, 'HENRY GARCIA', 97)
(11330, 'RYAN THOMPSON', 97)
(11223, 'HAILEY PATTERSON', 90)


### 4.4 Sazonalidade das vendas (por mês)

In [16]:
run("""
SELECT
  substr(OrderDate, 1, instr(OrderDate, '/') - 1) AS Mes,
  SUM(OrderQuantity) AS TotalVendida
FROM Sales_All
GROUP BY Mes
ORDER BY CAST(Mes AS INT);
""")

('1', 7446)
('2', 7260)
('3', 7791)
('4', 8174)
('5', 8734)
('6', 8784)
('7', 2201)
('8', 6236)
('9', 6166)
('10', 6610)
('11', 6523)
('12', 8249)


✅ Insights iniciais — ETAPA 1 (SQL Analytics)

A análise inicial do período entre 2015 e 2017 revela crescimento consistente no volume de vendas, saindo de 2.630 unidades em 2015 para 45.314 unidades em 2017, totalizando 84.174 unidades vendidas.
A análise de clientes indica uma base altamente distribuída: os 10 maiores clientes somam apenas 991 unidades, representando aproximadamente 1,18% do volume total. Isso sugere baixa dependência de grandes compradores e um modelo comercial orientado à escala.
Em termos de produtos, destaca‑se o Water Bottle - 30 oz. como carro‑chefe, com 7.967 unidades vendidas, reforçando a importância de itens de alta rotatividade.
A análise de sazonalidade demonstra picos de vendas entre abril e junho, além de um aumento significativo em dezembro, indicando comportamento típico de varejo com forte impacto sazonal.

# --------------------------------------------------------------------------------------------------------

# 🧑‍💼 Briefing do gerente (realista)

### “Danyllo, agora precisamos preparar os dados para o dashboard executivo.
### O BI não deve fazer tratamento pesado.
### Quero um dataset limpo, consolidado e padronizado.”

# Objetivo da Etapa 2

Ao final da ETAPA 2, vamos ter: 

✅ Uma tabela fato única

✅ Datas normalizadas (YYYY-MM-DD)

✅ Colunas derivadas prontas para BI

✅ Dataset limpo, confiável e documentado

✅ Dados prontos para consumo no Power BI

# 1. Ler os dados do SQLite

In [17]:
import sqlite3
import pandas as pd

# Conectar ao banco
conn = sqlite3.connect("adventureworks.db")

# Carregar tabelas
sales = pd.read_sql("SELECT * FROM Sales_All", conn)
products = pd.read_sql("SELECT * FROM Products", conn)
customers = pd.read_sql("SELECT * FROM Customers", conn)

conn.close()

## 1.1 Normalizar datas (regra críica de ETL)

Atualmente a data está assim:

**1/3/2015**

Isso NÃO pode ir para o Power BI desse jeito.

### Converter para formato ISO (YYYY-MM-DD):

In [18]:
sales['OrderDate'] = pd.to_datetime(sales['OrderDate'], format='%m/%d/%Y')

### Agora podemos criar colunas derivadas: 

In [19]:
sales['Year'] = sales['OrderDate'].dt.year
sales['Month'] = sales['OrderDate'].dt.month

## 1.2 Criar Dimensão cliente (nome completo)

No banco, nomes estavam separados.

Agora criamos isso no ETL, onde faz mais sentido.

In [20]:
customers['FullName'] = customers['FirstName'] + ' ' + customers['LastName']

## 1.3 Juntando tudo (criar fato consolidada)

Criando a **Fato Vendas**, como um Data Warehouse real. 

In [21]:
fact_sales = (
    sales
    .merge(products, on="ProductKey", how="left")
    .merge(customers[['CustomerKey','FullName']], on="CustomerKey", how="left")
)

### Agora temos:

* Data
* Produto
* Cliente
* Quantidade
* Tudo num único dataset.

## 1.4 Criando métricas de negócio 

#### Vamos criar um preço médio por produto: 

In [22]:
import numpy as np

np.random.seed(42)
fact_sales['UnitPrice'] = np.random.uniform(5, 100, size=len(fact_sales))

### Criar métricas:

In [23]:
fact_sales['TotalValue'] = fact_sales['OrderQuantity'] * fact_sales['UnitPrice']
fact_sales['Cost'] = fact_sales['TotalValue'] * np.random.uniform(0.5, 0.8)
fact_sales['GrossMargin'] = fact_sales['TotalValue'] - fact_sales['Cost']

### Isso permite: 

* Receita
* Custo
* Margem
* KPIs financeiros no Power BI

## 1.5 Selecionando colunas finais (pensando em BI)

In [28]:
final_dataset = fact_sales[
    [
        'OrderDate',
        'Year',
        'Month',
        'FullName',
        'ProductName',
        'OrderQuantity',
        'UnitPrice',
        'TotalValue',
        'Cost',
        'GrossMargin'
    ]
].copy()

## 1.6 Exportar para o Power BI

In [25]:
final_dataset.to_csv("fact_sales_powerbi.csv", index=False)

# ------------------------------------------------------------------------------------- 

# 🧑‍💼 Novo briefing do gerente (simulação real)

### “Danyllo, preciso de um painel executivo que eu possa usar em reuniões.

### Ele precisa responder em poucos segundos:

### * Como estão as vendas?
### * Estamos crescendo?
### * O que vende mais?
### * Quando vendemos mais?
### * Como está nossa margem?

### Nada técnico, tudo objetivo.”

# ETAPA 3 -> Limitação devido ao Power Bi Web

Devido esnosso relatório baseado em CSV e por questões de limitação e bloqueo, quando usamos o Power BI Web, não é permitido criar medidas DAX nesse tipo de dataset. Medidas DAX só podem ser criadas se o modelo vier do Power BI desktop. 

## Medida de contorno: 

➡️ Criar as métricas no Python, e no Power BI Web apenas visualizar.

👉 Vamos seguir com o projeto assim:

✅ Python:

ETL pesado ✅

Métricas criadas ✅

Dados BI‑ready ✅

✅ Power BI Web:

Visualizações ✅

KPIs por agregação ✅

Storytelling ✅

### Vamos seguir com 3 blocos, sendo eles: 

1️⃣ Quais métricas vamos deixar prontas no Python

2️⃣ Estruturar o CSV final para o Power BI Web

3️⃣ Montar o dashboard no Power BI Web sem DAX

## BLOCO 1 - Criando métricas

O princípio aqui é:

💡 Tudo que for cálculo → Python

💡 Tudo que for visual → Power BI Web

In [29]:
final_dataset.head(5)

,OrderDate,Year,Month,FullName,ProductName,OrderQuantity,UnitPrice,TotalValue,Cost,GrossMargin
0,2015-01-01,2015,1,JOHN THOMAS,"Road-650 Black, 58",1,40.581311,40.581311,25.012571,15.568740
1,2015-01-01,2015,1,KYLE WASHINGTON,"Road-150 Red, 48",1,95.317859,95.317859,58.749821,36.568039
2,2015-01-01,2015,1,ROSS SANZ,"Mountain-100 Black, 44",1,74.539424,74.539424,45.942889,28.596536
3,2015-01-01,2015,1,SETH LEWIS,"Road-650 Black, 44",1,61.872556,61.872556,38.135577,23.736979
4,2015-01-02,2015,1,ALEJANDRO CHEN,"Road-150 Red, 48",1,19.821771,19.821771,12.217285,7.604486


In [30]:
## ✅ Métrica 1 — Margem percentual

final_dataset['MarginPct'] = final_dataset['GrossMargin'] / final_dataset['TotalValue']

In [32]:
## ✅ Métrica 2 — Vendas agregadas por Ano (para KPIs e gráfico)

sales_by_year = (
    final_dataset
    .groupby('Year', as_index=False)
    .agg(
        TotalSales=('TotalValue', 'sum'),
        TotalQuantity=('OrderQuantity', 'sum'),
        TotalMargin=('GrossMargin', 'sum')  
    )
)

sales_by_year['MarginPct'] = sales_by_year['TotalMargin'] / sales_by_year['TotalSales']

In [33]:
## ✅ Métrica 3 — Top 10 Produtos

top_products = (
    final_dataset
    .groupby('ProductName', as_index=False)
    .agg(
        TotalQuantity=('OrderQuantity', 'sum'),
        TotalSales=('TotalValue', 'sum')
    
    )
    .sort_values('TotalQuantity', ascending=False)
    .head(10)
)

In [35]:
## ✅ Métrica 4 — Top 10 Clientes

top_customers = (
    final_dataset
    .groupby('FullName', as_index=False)
    .agg(
        TotalQuantity=('OrderQuantity', 'sum'),
        TotalSales=('TotalValue', 'sum')
    )
    .sort_values('TotalQuantity', ascending=False)
    .head(10)
)

In [36]:
## ✅ Métrica 5 — Sazonalidade (por mês) 

sales_by_month = (
    final_dataset
    .groupby('Month', as_index=False)
    .agg(
        TotalQuantity=('OrderQuantity', 'sum'),
        TotalSales=('TotalValue', 'sum')
    )
    .sort_values('Month')
)

## BLOCO 2 - Exportação dos arquivos CSV

In [39]:
final_dataset.to_csv("fact_sales_powerbi.csv", index=False)
sales_by_year.to_csv("kpi_sales_by_year.csv", index=False)
top_products.to_csv("top_products.csv", index=False)
top_customers.to_csv("top_customers.csv", index=False)
sales_by_month.to_csv("sales_by_month.csv", index=False)

## BLOCO 3 - Dashboard no Power BI WEB

📌 Importação
No Power BI Web:

Criar → Upload de arquivo

Subir todos os CSVs:

* fact_sales_powerbi.csv
* kpi_sales_by_year.csv
* top_products.csv
* top_customers.csv
* sales_by_month.csv

Página 1 - Visão Executiva
===================

Dados: kpi_sales_by_year.csv

Cards

* TotalSales (soma)
* TotalMargin (soma)
* MarginPct (média)
* TotalQuantity (soma)

Gráfico de colunas

* Eixo: Year
* Valor: TotalSales

👉 Essa página responde: “Estamos crescendo?”

Página 2 — PRODUTOS
=========================

Dados: top_products.csv

Gráfico de barras

* Eixo: ProductName
* Valor: TotalQuantity
* Ordenar desc

👉 Insight visual direto:

Produtos de alta rotatividade puxam o desempenho

Página 3 — CLIENTES
=============

Dados: top_customers.csv

Tabela ou barras

Cliente × TotalQuantity



👉 Reforça:

Baixa concentração de clientes (1,18%)

Página 4 — SAZONALIDADE
==

Dados: sales_by_month.csv

Gráfico de colunas

Month × TotalQuantity


👉 Picos claros:

Abril–Junho
Dezembro